server/src/Subscription.hs

```diff
@@ -64,22 +68,30 @@ unsubscribe path subid (SubscriptionTree here inner) =
 
 -- Invoke f for all subscribers to the path. The subscribers get passed the
 -- subvalue at the path that they are subscribed to.
-broadcast :: (Value -> v -> IO ()) -> [Text] -> Value -> SubscriptionTree k v -> IO ()
-broadcast f path value (SubscriptionTree here inner) =
+broadcast :: (v -> Value -> IO ()) -> [Text] -> Value -> SubscriptionTree k v -> IO ()
+broadcast f path value tree =
+  let notifications = execWriter $ broadcast' path value tree
+  in Async.mapConcurrently_ (uncurry f) notifications
+
+-- Like broadcast, but return a list of notifications rather than invoking an
+-- effect on each of them.
+broadcast' :: [Text] -> Value -> SubscriptionTree k v -> Writer [(v, Value)] ()
```

Comment:
Could you add some docs to explain the use of Writer here?
Is the main advantage that we have fewer values to pass around in recursive calls?

Maybe also mention that `v` is one websocket connection here. Or better just call it `connection` instead of `v`.
(It's nice to abstract this, for testing, but it makes it now also harder to follow along).

Reply:
Without it, recursion would be a bit trickier, we'd have to use `foldMap` and `(<>)`. With `Writer` we can use `do`-notation which is less obscure. I'd like to get rid of `Writer` from the type signature.